### PR TITLE
Transfer Server VPC Endpoint Type and User Home Directory Type / Mapping

### LINKED ISSUE #11281: aws_transfer_user should provide home-directory-mappings option.
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

AWS CLI https://docs.aws.amazon.com/cli/latest/reference/transfer/create-user.html provides` home-directory-type` and `home-directory-mappings` to control logical mapping between S3 and SFTP. Terraform AWS provider does not support the feature.

### New or Affected Resource(s)

* aws_transfer_user

### Potential Terraform Configuration

```hcl
resource "aws_transfer_user" "sftp_user_dbs" {
  # other options
  # ...
  home_directory_type = "LOGICAL"
  home_directory_mapping = {
    entry = "your-personal-report.pdf"
    target = "/bucket3/customized-reports/${transfer:UserName}.pdf" 
  }
}
```

### References

* https://docs.aws.amazon.com/cli/latest/reference/transfer/create-user.html

### LINKED ISSUE #11569: aws_transfer_server: support Elastic IPs
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description
AWS Transfer now supports Elastic IPs. With this, there's no more need to create a NLB to expose the server in known IP address.

Terraform provider should support these as well.

### New or Affected Resource(s)
* `aws_transfer_server`

### Potential Terraform Configuration

```hcl
resource "aws_transfer_server" "foo" {
  endpoint_details {
    address_allocation_ids = ["eipalloc-..."]
    subnet_ids = ["subnet-..."]
    vpc_id = "vpc-..."
    vpc_endpoint_id = "vpce-..."
  }
  endpoint_type = "VPC"
}
```

### References
Announcement: https://aws.amazon.com/about-aws/whats-new/2020/01/aws-transfer-for-sftp-supports-vpc-security-groups-and-elastic-ip-addresses/
Docs: https://docs.aws.amazon.com/transfer/latest/userguide/create-sftp-server-in-vpc.html
Requires [AWS SDK v1.28.0](https://github.com/aws/aws-sdk-go/releases/tag/v1.28.0):
* #11501

### LINKED ISSUE #11593: AWS Sftp now support elastic ips
_This issue was originally opened by @KVInventoR as hashicorp/terraform#23851. It was migrated here as a result of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._
    
<hr>
    
Hi All

A lot of time ago, AWS announced supports security groups and elastic ip addresses for sftp server:
https://aws.amazon.com/about-aws/whats-new/2020/01/aws-transfer-for-sftp-supports-vpc-security-groups-and-elastic-ip-addresses/

do you have a plans to support it in aws module?

According to documentation: 
https://docs.aws.amazon.com/cli/latest/reference/transfer/create-server.html

it's possible to set different options:
```--endpoint-type (string)

The type of VPC endpoint that you want your SFTP server to connect to. You can choose to connect to the public internet or a virtual private cloud (VPC) endpoint. With a VPC endpoint, you can restrict access to your SFTP server and resources only within your VPC.

Possible values:

PUBLIC
VPC
VPC_ENDPOINT
```

but yours:
https://www.terraform.io/docs/providers/aws/r/transfer_server.html
supports only PUBLIC
```
endpoint_type - (Optional) The type of endpoint that you want your SFTP server connect to. If you connect to a VPC_ENDPOINT, your SFTP server isn't accessible over the public internet. If you want to connect your SFTP server via public internet, set PUBLIC. Defaults to PUBLIC.

```

any plans to update sftp server?

### LINKED ISSUE #11632: aws_transfer_user: Support "restricted" home directory
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

AWS Transfer for SFTP supports restricting a user's home directory:

"When a home directory is set to "Restricted", your users will not be able to access anything outside of that folder, nor will they be able to see the S3 bucket or folder name."

Screenshot attached.

Screen Shot 2020-01-16 at 4 49 08 PM

I couldn't find a way to do this via the Terraform resource (only could restrict access via a policy), but would be helpful to have available for setting up new accounts.

### New or Affected Resource(s)

* aws_transfer_user

### Potential Terraform Configuration

```hcl
resource "aws_transfer_user" "testuser" {
  server_id = "${aws_transfer_server.main.id}"
  user_name = "testuser"
  role      = "${aws_iam_role.default_access.arn}"
  home_directory = "/${aws_s3_bucket.default.bucket}"
  restricted = true

  tags = {
    NAME = "testuser"
  }
}
```

### LINKED ISSUE #11724: aws_transfer_server does not currently support endpoint_type: "VPC"
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

```
Terraform v0.12.20
+ provider.aws v2.44.0
```

### Affected Resource(s)

* aws_transfer_server

### Terraform Configuration Files

```hcl
resource "aws_transfer_server" "main" {
  identity_provider_type = "SERVICE_MANAGED"
  logging_role                 =  aws_iam_role.transfer.arn
  endpoint_type              = "VPC"

  tags = {
    NAME = "transfer-server"
  }

  endpoint_details {
    vpc_endpoint_id = aws_vpc_endpoint.transfer.id
  }
}

```

### Debug Output

```
Error: expected endpoint_type to be one of [PUBLIC VPC_ENDPOINT], got VPC
```

### Expected Behavior

Updated endpoint type to VPC.

### Actual Behavior

Produced error above: `Error: expected endpoint_type to be one of [PUBLIC VPC_ENDPOINT], got VPC`

### Steps to Reproduce

1. Create publicly-available Transfer server (& terraform apply):
```
resource "aws_iam_role" "transfer" {
  name = "transfer-server-iam-role"

  assume_role_policy = <<EOF
{
    "Version": "2012-10-17",
    "Statement": [
        {
        "Effect": "Allow",
        "Principal": {
            "Service": "transfer.amazonaws.com"
        },
        "Action": "sts:AssumeRole"
        }
    ]
}
EOF
}

resource "aws_transfer_server" "main" {
  identity_provider_type = "SERVICE_MANAGED"
  logging_role           = "${aws_iam_role.transfer.arn}"
  endpoint_type          = "PUBLIC"

  tags = {
    NAME = "transfer-server"
  }
}
```
2. Create dependencies for the changes we're about to make in the AWS console (& terraform apply):

```hcl
variable "aws_az1" {
  description = "AWS availability zone 1"
  default     = "us-east-1a"
}

variable "aws_az2" {
  description = "AWS availability zone 2"
  default     = "us-east-1b"
}

resource "aws_vpc" "main" {
  cidr_block = "10.0.0.0/16"
  enable_dns_hostnames = true
}

resource "aws_subnet" "public" {
  vpc_id = "${aws_vpc.main.id}"
  cidr_block = "10.0.0.0/16"
  availability_zone = "${var.aws_az2}"

  tags = {
    name = "cmr-datashare-public"
  }
}

resource "aws_internet_gateway" "default" {
  vpc_id = "${aws_vpc.main.id}"
}

resource "aws_route_table" "internet_access" {
  vpc_id = "${aws_vpc.main.id}"

  route {
    cidr_block = "0.0.0.0/0"
    gateway_id = "${aws_internet_gateway.default.id}"
  }
}

resource "aws_route_table_association" "subnet_route_table" {
  subnet_id = "${aws_subnet.public.id}"
  route_table_id = "${aws_route_table.internet_access.id}"
}

resource  "aws_eip" "sftp" {
  vpc = true
  depends_on = [
    "aws_internet_gateway.default"
  ]
}

# Stays the same:
resource "aws_iam_role" "transfer" {
  name = "transfer-server-iam-role"

  assume_role_policy = <<EOF
{
    "Version": "2012-10-17",
    "Statement": [
        {
        "Effect": "Allow",
        "Principal": {
            "Service": "transfer.amazonaws.com"
        },
        "Action": "sts:AssumeRole"
        }
    ]
}
EOF
}

resource "aws_transfer_server" "main" {
  identity_provider_type = "SERVICE_MANAGED"
  logging_role           = "${aws_iam_role.transfer.arn}"
  endpoint_type          = "PUBLIC"

  tags = {
    NAME = "transfer-server"
  }
}
```
3. In AWS console, visit your Transfer server. Stop instance and click "edit", change to be of type "VPC". Associate the resources you made in the previous step.

Screen Shot 2020-01-22 at 4 13 09 PM

4. Save changes to server and start again.

5. If you now `terraform apply` again, you'll see that the endpoint type for the `aws_transfer_server` wants to change as follows: `~ endpoint_type          = "VPC" -> "PUBLIC"`. If you try to change the terraform config to `endpoint_type = "VPC"` then you get the following error: `Error: expected endpoint_type to be one of [PUBLIC VPC_ENDPOINT], got VPC`.

### Important Factoids

It seems like AWS is encouraging people to use the "VPC" endpoint instead of the "VPC_ENDPOINT" endpoint type in the AWS Console and the docs (see reference below). The VPC_ENDPOINT is not available as an option if you create the Transfer server from scratch in the AWS console.

### References

* [AWS Docs: Creating an Internet-Facing Endpoint for Your SFTP Server](https://docs.aws.amazon.com/transfer/latest/userguide/create-sftp-server-in-vpc.html#create-internet-facing-endpoint)
* [AWS Docs: Moving away from VPC_ENDPOINT](https://docs.aws.amazon.com/transfer/latest/userguide/create-sftp-server-in-vpc.html#change-sftp-server-endpoint-type): "If you have an existing AWS SFTP server in a VPC displayed as VPC_ENDPOINT, we recommend that you modify it to the new VPC endpoint type. With this new endpoint type, you no longer need to use a Network Load Balancer (NLB) to associate Elastic IP addresses with your server's endpoint. Also, you can use VPC security groups to restrict access to your server's endpoint. However, you can continue to use the VPC_ENDPOINT endpoint type as needed."

### PR DESCRIPTION
**Pull request by sshearn**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #11281
Closes #11632
Closes #11724
Closes #11569
Closes #11593

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note

Transfer server now supports VPC for endpoint_type.
```

Output from acceptance testing:

```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSTransferServer_Vpc'

...
```

```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSTransferUser_restricted'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSTransferUser_restricted -timeout 120m
=== RUN   TestAccAWSTransferUser_restricted
=== PAUSE TestAccAWSTransferUser_restricted
=== CONT  TestAccAWSTransferUser_restricted
--- PASS: TestAccAWSTransferUser_restricted (40.44s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       41.599s
```

Note:

`TestAccAWSTransferServer_Vpc` test failed because it couldn't release the eip's on destroy due to policies on my account. But the actual test works.

### CODE REVIEW

**Comment by sshearn** [comment]
@bflad I'm using my fork of this in prod. So I know it works great. I'm hoping this can get into the next release so I don't have to maintain my fork for long though. Let me know if I can help with anything!

**Comment by momer** [comment]
@sshearn at a quick glance, this looks great. Watching this PR as well! Thank you for this.

**Comment by momer** [comment]
@sshearn This is great - I played around quite a bit, here are a couple of pieces of feedback:

1. I'd add a note about `home_directory` not being able to be set if `home_directory_mappings` is set.
2. For a user with `home_directory_mappings`, if a change-set is applied which indicates a change of the `Target` will occur, and is then applied, it will succeed, but no change will actually occur (*i.e.* the mapping does not actually change).

All in all, this was a joy to use. Thank you for this contribution!

File: aws/resource_aws_transfer_server.go [hunk 1]
```diff
@@ -52,8 +54,9 @@ func resourceAwsTransferServer() *schema.Resource {
 				Elem: &schema.Resource{
 					Schema: map[string]*schema.Schema{
 						"vpc_endpoint_id": {
-							Type:     schema.TypeString,
-							Required: true,
+							Type:          schema.TypeString,
+							Optional:      true,
+							ConflictsWith: []string{"endpoint_details.0.address_allocation_ids", "endpoint_details.0.subnet_ids", "endpoint_details.0.vpc_id"},
 							ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
```
Comment (ewbankkit) [inline aws/resource_aws_transfer_server.go:60]: I suggest removing this `ValidateFunc`, we don't normally check for resource ID formatting. Thanks.

File: aws/resource_aws_transfer_server.go [hunk 2]
```diff
@@ -64,11 +67,45 @@ func resourceAwsTransferServer() *schema.Resource {
...
+							Type:          schema.TypeString,
+							Optional:      true,
+							ValidateFunc:  validation.NoZeroValues,
+							ConflictsWith: []string{"endpoint_details.0.vpc_endpoint_id"},
 						},
 					},
 				},
 			},
 
+			"vpce_security_group_ids": {
```
Comment (ewbankkit) [inline aws/resource_aws_transfer_server.go:101]: Rename to `vpc_endpoint_security_group_ids`.

**Comment by ewbankkit** [comment]
@sshearn Thanks for this.
We usually prefer not to manipulate AWS resources from differing AWS services in a single Terraform resource and I notice that the `vpce_security_group_ids` attribute manipulates an EC2 VPC Endpoint resource. Is there a way of managing these security groups through a [`aws_vpc_endpoint`](https://www.terraform.io/docs/providers/aws/r/vpc_endpoint.html) resource?

**Comment by derekmurawsky** [comment]
> @sshearn Thanks for this.
> We usually prefer not to manipulate AWS resources from differing AWS services in a single Terraform resource and I notice that the `vpce_security_group_ids` attribute manipulates an EC2 VPC Endpoint resource.
> Is there a way of managing these security groups through a [`aws_vpc_endpoint`](https://www.terraform.io/docs/providers/aws/r/vpc_endpoint.html) resource?
> I remember [this comment](https://github.com/terraform-providers/terraform-provider-aws/pull/2253#issuecomment-345554521) about why there is no VPC Endpoint/Security Group Association resource.

The transfer server brings up its own VPC endpoint that cannot be brought up or configured via the normal endpoint service. 
I manually brought up the transfer server and this is the endpoint that was created (top one in the image). 
image

When creating an endpoint via the normal method, there is no way to create an endpoint of this type. 
image

As such, I would suggest the proper way to manage this service-created endpoint would be through the transfer-server module.

**Comment by derekmurawsky** [comment]
> As such, I would suggest the proper way to manage this service-created endpoint would be through the transfer-server module.

I did some more digging today into this. I created a transfer server and then attempted to use the following code to attach the security groups per @ewbankkit 's suggestion.

```
data "aws_transfer_server" "transfer-server" {
  server_id = aws_cloudformation_stack.transfer-server.outputs["ServerId"]
}

data "aws_vpc_endpoint" "transfer-server" {
  vpc_id = aws_vpc.external.id
}

resource "aws_network_interface_sg_attachment" "transfer-server-endpoint" {
  for_each             = data.aws_vpc_endpoint.transfer-server.network_interface_ids
  security_group_id    = aws_security_group.allow_sftp_all.id
  network_interface_id = each.value
}
```

It failed with the following

``` 
Error: AuthFailure: You do not have permission to access the specified resource.
        status code: 400, request id: 4be36bc7-ffff-4e11-af1a-977a52646147

  on transfer-server-cf.tf line 20, in resource "aws_network_interface_sg_attachment" "transfer-server-endpoint":
  20: resource "aws_network_interface_sg_attachment" "transfer-server-endpoint" {

```

This error makes sense; these interfaces are managed through the endpoint service and, as such, can't be managed directly. The web UI also has the same error: 
image

There doesn't appear to be a way to manage this without something like an aws_**vpc_endpoint**_sg_attachement resource.

**Comment by ewbankkit** [comment]
@derekmurawsky Thanks for looking into this. We are considering creating an `aws_vpc_endpoint_security_group_association` resource to handle this functionality. I'll keep this issue update with the status.

**Comment by sshearn** [comment]
> @derekmurawsky Thanks for looking into this. We are considering creating an `aws_vpc_endpoint_security_group_association` resource to handle this functionality. I'll keep this issue update with the status.

I can update this PR to account for this once it's merged in.

File: aws/resource_aws_transfer_server.go [hunk 3]
```diff
@@ -392,19 +454,203 @@ func expandTransferServerEndpointDetails(l []interface{}) *transfer.EndpointDeta
...
 func flattenTransferServerEndpointDetails(endpointDetails *transfer.EndpointDetails) []interface{} {
 	if endpointDetails == nil {
 		return []interface{}{}
 	}
 
-	e := map[string]interface{}{
-		"vpc_endpoint_id": aws.StringValue(endpointDetails.VpcEndpointId),
+	e := make(map[string]interface{})
+	if endpointDetails.VpcEndpointId != nil {
+		e["vpc_endpoint_id"] = *endpointDetails.VpcEndpointId
```
Comment (DrFaust92) [inline aws/resource_aws_transfer_server.go:485]: use sdk wrapper `aws.StringValue()` instead of `*`

File: aws/resource_aws_transfer_server.go [hunk 4]
```diff
@@ -392,19 +454,203 @@ func expandTransferServerEndpointDetails(l []interface{}) *transfer.EndpointDeta
...
+	}
+	if endpointDetails.AddressAllocationIds != nil {
+		e["address_allocation_ids"] = endpointDetails.AddressAllocationIds
+	}
+	if endpointDetails.SubnetIds != nil {
+		e["subnet_ids"] = endpointDetails.SubnetIds
+	}
+	if endpointDetails.VpcId != nil {
+		e["vpc_id"] = *endpointDetails.VpcId
```
Comment (DrFaust92) [inline aws/resource_aws_transfer_server.go:494]: use sdk wrapper `aws.StringValue()` instead of `*`

File: aws/resource_aws_transfer_server.go [hunk 5]
```diff
@@ -392,19 +454,203 @@ func expandTransferServerEndpointDetails(l []interface{}) *transfer.EndpointDeta
...
 	}
 
-	e := map[string]interface{}{
-		"vpc_endpoint_id": aws.StringValue(endpointDetails.VpcEndpointId),
+	e := make(map[string]interface{})
+	if endpointDetails.VpcEndpointId != nil {
+		e["vpc_endpoint_id"] = *endpointDetails.VpcEndpointId
+	}
+	if endpointDetails.AddressAllocationIds != nil {
+		e["address_allocation_ids"] = endpointDetails.AddressAllocationIds
```
Comment (DrFaust92) [inline aws/resource_aws_transfer_server.go:488]: wrap value with `flattenStringSet()`

File: aws/resource_aws_transfer_server.go [hunk 6]
```diff
@@ -392,19 +454,203 @@ func expandTransferServerEndpointDetails(l []interface{}) *transfer.EndpointDeta
...
+	}
+	if endpointDetails.SubnetIds != nil {
+		e["subnet_ids"] = endpointDetails.SubnetIds
```
Comment (DrFaust92) [inline aws/resource_aws_transfer_server.go:491]: wrap value with `flattenStringSet()`

File: aws/resource_aws_transfer_user.go [hunk 7]
```diff
@@ -261,3 +311,44 @@ func waitForTransferUserDeletion(conn *transfer.Transfer, serverID, userName str
...
+	return ms
+}
+
+func flattenTransferServerUserHomeDirectoryMappings(m []*transfer.HomeDirectoryMapEntry) []map[string]interface{} {
+	ms := make([]map[string]interface{}, 0)
+
+	for _, e := range m {
+		if e.Entry != nil && e.Target != nil {
+			m := make(map[string]interface{})
+			m["entry"] = *e.Entry
```
Comment (DrFaust92) [inline aws/resource_aws_transfer_user.go:342]: use sdk wrapper`aws.StringValue()`

File: aws/resource_aws_transfer_user.go [hunk 8]
```diff
@@ -261,3 +311,44 @@ func waitForTransferUserDeletion(conn *transfer.Transfer, serverID, userName str
...
+			m["target"] = *e.Target
```
Comment (DrFaust92) [inline aws/resource_aws_transfer_user.go:343]: same as above

File: aws/resource_aws_transfer_server_test.go [hunk 9]
```diff
@@ -111,6 +111,51 @@ func TestAccAWSTransferServer_basic(t *testing.T) {
 	})
 }
 
+func TestAccAWSTransferServer_Vpc(t *testing.T) {
+	var conf transfer.DescribedServer
+
```
Comment (DrFaust92) [inline aws/resource_aws_transfer_server_test.go:116]: add `resourceName := "aws_transfer_server.test"` and replace all `"aws_transfer_server.test"` occurrences

File: aws/resource_aws_transfer_user_test.go [hunk 10]
```diff
@@ -43,6 +43,38 @@ func TestAccAWSTransferUser_basic(t *testing.T) {
...
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:      func() { testAccPreCheck(t); testAccPreCheckAWSTransfer(t) },
+		IDRefreshName: "aws_transfer_user.foo",
+		Providers:     testAccProviders,
+		CheckDestroy:  testAccCheckAWSTransferUserDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccAWSTransferUserConfig_restricted(rName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSTransferUserExists("aws_transfer_user.foo", &conf),
```
Comment (DrFaust92) [inline aws/resource_aws_transfer_user_test.go:59]: add resourceName := "aws_transfer_user.foo" and replace all "aws_transfer_user.foo" occurrences

**Comment by dturnbu** [comment]
Is anyone planning on making the changes suggested by @DrFaust92? 

I'd really like to see this PR merged soon....

File: aws/resource_aws_transfer_user.go [hunk 11]
```diff
@@ -32,14 +32,46 @@ func resourceAwsTransferUser() *schema.Resource {
...
+							ValidateFunc: validation.StringLenBetween(1, 1024),
+						},
+					},
+				},
 			},
 
 			"policy": {
 				Type:             schema.TypeString,
 				Optional:         true,
+				ConflictsWith:    []string{"home_directory_mappings"},
```
Comment (s33dunda) [inline aws/resource_aws_transfer_user.go:74]: This is not a true conflict. AWS SDK Allows for policy along with the home directories.

**Comment by dturnbu** [comment]
oh boy, it sure would be swell if this PR could get merged..............

**Comment by mjhoffman65** [comment]
Anything I can help with this?

**Comment by andwoods** [comment]
any update on the VPC endpoint type PR? I could use this functionality

**Comment by epang-dev** [comment]
> any update on the VPC endpoint type PR? I could use this functionality

Same here, I would also like to use this functionality.

**Comment by asimonrose** [comment]
Yes - we’re very keen as well, any indication of availability of this would really help.

**Comment by erikorange** [comment]
Any update on availability?  I'd really like to avoid using CloudFormation to somehow deploy this VPC endpoint.  Thanks in advance.

**Comment by bflad** [comment]
Hi @sshearn 👋  Thank you for your work here. Are you still interested in working on this contribution? There are some review items noted above that prevent merging this and we separately just merged `aws_transfer_user` resource changes that preceded this contribution, so these should be removed from this pull request to remove the merge conflicts. Please reach out if you have any questions.

If we do not hear back in two weeks, we will default to providing review of #15105, which includes these commits and some of these review item fixes.

**Comment by sshearn** [comment]
> Hi @sshearn 👋 Thank you for your work here. Are you still interested in working on this contribution? There are some review items noted above that prevent merging this and we separately just merged `aws_transfer_user` resource changes that preceded this contribution, so these should be removed from this pull request to remove the merge conflicts. Please reach out if you have any questions.
> 
> If we do not hear back in two weeks, we will default to providing review of #15105, which includes these commits and some of these review item fixes.

I basically abandoned this for my own provider that I've been using in production for months. I guess I could pick this back up and making the new required changes. I'll start working on it.

File: aws/resource_aws_transfer_server.go (hunk 1 above)
Comment (sshearn) [reply aws/resource_aws_transfer_server.go]: done

File: aws/resource_aws_transfer_server.go (hunk 5 above)
Comment (sshearn) [reply aws/resource_aws_transfer_server.go]: done

File: aws/resource_aws_transfer_server.go (hunk 6 above)
Comment (sshearn) [reply aws/resource_aws_transfer_server.go]: done

**Comment by sshearn** [comment]
All changes requested are addressed

File: website/docs/r/transfer_user.html.markdown [hunk 12]
```diff
@@ -77,9 +77,12 @@ The following arguments are supported:
...
 * `home_directory_type` - (Optional) The type of landing directory (folder) you mapped for your users' home directory. Valid values are `PATH` and `LOGICAL`.
 * `policy` - (Optional) An IAM JSON policy document that scopes down user access to portions of their Amazon S3 bucket. IAM variables you can use inside this policy include `${Transfer:UserName}`, `${Transfer:HomeDirectory}`, and `${Transfer:HomeBucket}`. Since the IAM variable syntax matches Terraform's interpolation syntax, they must be escaped inside Terraform configuration strings (`$${Transfer:UserName}`).  These are evaluated on-the-fly when navigating the bucket.
+* `home_directory_type` (Optional) The type of landing directory (folder) you want your users' home directory to be when they log into the SFTP server. If you set it to PATH, the user will see the absolute Amazon S3 bucket paths as is in their SFTP clients. If you set it `LOGICAL`, you will need to provide mappings in the `home_directory_mappings` for how you want to make S3 paths visible to your user. Allowed Values: `LOGICAL` | `PATH`.
```
Comment (bflad) [inline website/docs/r/transfer_user.html.markdown:80]: Thank you for those updates! Looks like there's still some lingering `aws_transfer_user` bits here (in this case, duplicate documentation being added)

Comment (sshearn) [reply website/docs/r/transfer_user.html.markdown]: done

**Comment by bflad** [comment]
Thank you for your work on this, @sshearn. I think this is now in a good enough state where we can discuss the overall approach here to prepare to get this functionality merged with a few relevant changes. 👍  Please reach out if you have any questions or if you would prefer to not spend additional time on this (no big deal either which way, we'll make sure you get credit).

1. Terraform AWS Provider resources should avoid cross-service functionality (e.g. no `service/ec2` handling in `aws_transfer_server` resource) as this type of handling has created many issues in the past and can introduce confusion about IAM permissions necessary for the resource. Our preference will be to create some form of Terraform AWS Provider functionality, after this pull request and separate from the `aws_transfer_server` resource, that manages the EC2 VPC Endpoint that is automatically created by the Transfer service. To that end and especially because it is not part of the Transfer API, let's remove all the `vpce_security_group_ids` and associated logic.
1. Switch the `vpc_endpoint_id` attribute from `DiffSuppressFunc` to `Computed: true` -- this will allow Terraform core to properly know that the value is unknown for downstream references when unconfigured.
1. Include a quick comment about why `AddressAllocationIds` needs to be handled via update after creation (e.g. reference https://docs.aws.amazon.com/transfer/latest/userguide/API_EndpointDetails.html)

Thanks again.

**Comment by andwoods** [comment]
@sshearn Thank you for your effort on this functionality! It is making a sizable impact for a lot of people!

**Comment by sshearn** [comment]
> Thank you for your work on this, @sshearn. I think this is now in a good enough state where we can discuss the overall approach here to prepare to get this functionality merged with a few relevant changes. 👍 Please reach out if you have any questions or if you would prefer to not spend additional time on this (no big deal either which way, we'll make sure you get credit).
> 
>     1. Terraform AWS Provider resources should avoid cross-service functionality (e.g. no `service/ec2` handling in `aws_transfer_server` resource) as this type of handling has created many issues in the past and can introduce confusion about IAM permissions necessary for the resource. Our preference will be to create some form of Terraform AWS Provider functionality, after this pull request and separate from the `aws_transfer_server` resource, that manages the EC2 VPC Endpoint that is automatically created by the Transfer service. To that end and especially because it is not part of the Transfer API, let's remove all the `vpce_security_group_ids` and associated logic.
> 
>     2. Switch the `vpc_endpoint_id` attribute from `DiffSuppressFunc` to `Computed: true` -- this will allow Terraform core to properly know that the value is unknown for downstream references when unconfigured.
> 
>     3. Include a quick comment about why `AddressAllocationIds` needs to be handled via update after creation (e.g. reference https://docs.aws.amazon.com/transfer/latest/userguide/API_EndpointDetails.html)
> 
> 
> Thanks again.

1. As I said before, this is addressed in #13737 which isn't merged yet. I will remove `service/ec2` references but that means that #13737 is a dependency for this to be valuable.
2. Will do.
3. The api only supported `AddressAllocationIds` on `UpdateServer` only. I can check if that's changed since I originally created this PR, and if its now supported on create, I can refactor that. It's referenced here: https://docs.aws.amazon.com/sdk-for-go/api/service/transfer/#EndpointDetails

**Comment by sshearn** [comment]
I tested via the cli:

```
% aws transfer create-server --endpoint-details '{"AddressAllocationIds":["eipalloc-XXX"],"SubnetIds":["subnet-XXX"],"VpcId":"vpc-XXX"}' --endpoint-type VPC

An error occurred (InvalidRequestException) when calling the CreateServer operation: AddressAllocationIds cannot be set in CreateServer
```

**Comment by sshearn** [comment]
I've removed `vpce_security_group_ids`, but had to leave in the support for waiting on the creation of the VPC endpoint in order to stop and then update the `AddressAllocationIds`.

**Review by bflad — APPROVED** [review]
Alrighty! After diving into this more, was able to fully remove the EC2 dependency by introducing retry logic for the specific error on update after creation: `ConflictException: VPC Endpoint state is not yet available` -- this allows us to not require unexpected IAM permissions for this resource and prevents any other potential cross-service eventual consistency that might occur. It was certainly interesting that the Transfer Server state of `ONLINE` didn't wait for that.

After fixing up the test configuration to remove the non-existent data source reference and fixing the `expandTransferServerEndpointDetails` function so VPC_ENDPOINT support continued to work, looks good, thank you @sshearn 🚀 

Output from acceptance testing:

```
--- PASS: TestAccAWSTransferServer_apigateway (190.91s)
--- PASS: TestAccAWSTransferServer_basic (196.10s)
--- PASS: TestAccAWSTransferServer_disappears (180.81s)
--- PASS: TestAccAWSTransferServer_forcedestroy (155.75s)
--- PASS: TestAccAWSTransferServer_hostKey (155.02s)
--- PASS: TestAccAWSTransferServer_Vpc (288.89s)
--- PASS: TestAccAWSTransferServer_vpcEndpointId (143.93s)
```

**Comment by nauxliu** [comment]
Why vpc_id and vpc_endpoint_id are conflicts with each other. 

I imported an existing sftp server and removed the `vpc_id` attribute from the `endpoint_details` following the error message, the tf plan show the VPC id will be removed.
image

The AWS Dashboard is showing both vpc_id and vpc_endpoint_id in the Endpoint details. doesn't that mean they can exist together?
image

**Comment by rowmow** [comment]
Thanks so much @sshearn for these additions. Is there a way to also manage a "custom hostname" within the endpoint_details additions? I did not see this mentions in the latest documentation.

https://docs.aws.amazon.com/transfer/latest/userguide/requirements-dns.html

**Comment by andwoods** [comment]
the documentation and examples do not show how to set Access type (internal or internet-facing) for endpoint type = VPC. How do we do this?

**Pull request merged**